finetune/util/group_metrics.py

```diff
@@ -0,0 +1,522 @@
+            if start < end and pred_entity["label"] == label_entity["label"]:
+                return Counts(1, 0, 0)
+    return Counts(0, 1, 1)
+
+def joint_superset_counts(pred, label):
+    """
+    Return TP if all label entities appear within pred entities
+
+    :param pred, label: A joint group, represeted as a dict
+    :return: A Counts namedtuple with TP, FP and FN counts
+    """
+    if (pred["label"] != label["label"]):
+        return Counts(0, 1, 1)
+    for label_entity in label["entities"]:
+        for pred_entity in pred["entities"]:
+            if (pred_entity["start"] <= label_entity["start"] and
+                pred_entity["end"] >= label_entity["end"] and
+                pred_entity["label"] == label_entity["label"]):
+                break
+        else:
+            return Counts(0, 1, 1)
+    return Counts(1, 0, 0)
+
+def get_count_fn(metric_type, span_type):
+    """
+    Returns a function used to count TPs, FPs and FNs between two groups
+
+    Group count functions compare based on the label of the group and the text
+    spans contained within the group.
+
+    Joint count functions compare based on the label of the group and the
+    entity spans contained within the group. Requires groups to be in joint
+    group format with an "entities" key.
+
+    :param metric_type: A string, either "group" or "joint"
+    :param span_type: A string, one of "exact", "token", "overlap", "superset"
+    :return: A function that takes two groups and returns a Counts namedtuple
+    """
+    fns = {
+        "group": {
+            "exact": group_exact_counts,
+            "token": group_token_counts,
+            "overlap": group_overlap_counts,
+            "superset": group_superset_counts,
+        },
+        "joint": {
+            "exact": joint_exact_counts,
+            "token": joint_token_counts,
+            "overlap": joint_overlap_counts,
+            "superset": joint_superset_counts,
+        },
+    }
+    return fns[metric_type.lower()][span_type.lower()]
+
+def calc_group_assignment(preds, labels, count_fn):
+    """
+    Calculate optimal prediction to ground truth group assignments
+
+    :param preds, labels: A list of regular or joint groups
+    :return assign_idxs: A list of tuples representing the mapping of predicted
+    groups to ground truth groups, where (i, j) means preds[i] maps to labels[j]
+    :return counts: A list of shape [len(preds), len(labels)], where
+    counts[i][j] contains the TP, FP, FN counts if predicted group i was
+    assigned to ground truth group j
+    """
+    counts = [[(0, 0, 0) for _ in len(labels)] for _ in len(preds)]
+    costs = [[0 for _ in len(labels)] for _ in len(preds)]
+
+    for i, pred in enumerate(preds):
+        for j, label in enumerate(labels):
+            TP, FP, FN = count_fn(pred, label)
+            counts[i][j] = (TP, FP, FN)
+            costs[i][j] =  -(TP - (FP + FN) / 2)
+
+    pred_idxs, label_idxs = linear_sum_assignment(costs)
```

Comment:
I don't think this quite works.  If costs.shape[0] < costs.shape[1], I don't think the counts won't be the desired TP / FP / FN counts.  You can maybe force the costs / counts matrix to be square and fill the rest of the matrix with FP (in case of len(preds) > len(labels)) or FN (in case of len(preds) < len(labels))?  Personally also think it makes sense to first group by label before running these calcs?

Reply:
Take this with a grain of salt because I need to stare at longer